src/org/sosy_lab/java_smt/solvers/z3/Z3FormulaCreator.java

```diff
@@ -931,8 +969,17 @@ public void forceClose() {
 
     // Force clean all ASTs, even those which were not GC'd yet.
     // Is a no-op if phantom reference handling is not enabled.
-    for (long ast : referenceMap.values()) {
-      Native.decRef(getEnv(), ast);
+    if (usePhantomReferences) {
+      Z3AstReference cur = referenceListHead.next;
+      assert cur != null;
+      while (cur.next != null) {
+        Native.decRef(environment, cur.z3Ast);
+        cur = cur.next;
+      }
+      Z3AstReference tail = cur;
+      // Bulk delete everything between head and tail
+      referenceListHead.next = tail;
+      tail.prev = referenceListHead;
```

Comment:
This doesn't really change anything, right?

All the `Z3AstReference` deleted here will either be already in the `ReferenceQueue` or will be added to it in the future, unless the whole context is freed before. So in no case the reference through the list should matter.

Deleting here produces an inconsistent doubly-linked list, but this should also not matter because `cleanupReferences` should never be called again.

Not sure whether I find it cleaner to keep the deletion or to remove it.

(If it is removed, the `while` loop can be rewritten with `for`.)

Reply:
> All the `Z3AstReference` deleted here will either be already in the `ReferenceQueue` or will be added to it in the future, unless the whole context is freed before. So in no case the reference through the list should matter.

No, `PhantomReferences` that are not referenced anymore are themselves eligible for garbage collection and will never get enqueued because the GC won't find them anymore!
Otherwise we would not even have to bother with storing the phantom references in the first place.

It is true though that as long as there is a single reference enqueued, the whole list is reachable through that single reference and hence cannot be GC'd early. This also means that all references will eventually get enqueued...

Then I suggest to bulk delete the list, and then(*) poll the `ReferenceQueue` until it is empty.
This will guarantee that all phantom references can be GC'd and no reference will ever get enqueued again.

> 
> Deleting here produces an inconsistent doubly-linked list, but this should also not matter because `cleanupReferences` should never be called again.
> 
It is true that`cleanupReferences` should not be called again. Even if the list was kept consistent, this would still cause double decrements of the native reference counters and hence be incorrect (this behaviour matches the previous one). If I clear the queue as suggested above, this problem will also be gone! Furthmore the list is guaranteed to be fully consistent afterwards!

(*) The ordering is important. If one clears the queue first, it may get refilled during the deletion of the list